SmartDeviceLink/SDLFileManager.m

```diff
@@ -209,16 +210,126 @@ - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable
                                                                                  }
 
                                                                                  if (handler != nil) {
-                                                                                     handler(YES, self.bytesAvailable, nil);
+                                                                                     handler(success, self.bytesAvailable, error);
                                                                                  }
                                                                              }];
 
     [self.transactionQueue addOperation:deleteOperation];
 }
 
+- (void)deleteRemoteFilesWithNames:(NSArray<SDLFileName *> *)names completionHandler:(nullable SDLFileManagerMultiDeleteCompletionHandler)completionHandler {
+    if (names.count == 0) {
+        return completionHandler([NSError sdl_fileManager_noFilesError]);
+    }
+
+    NSMutableDictionary *failedDeletes = [[NSMutableDictionary alloc] init];
+
+    dispatch_group_t deleteFilesTask = dispatch_group_create();
+    dispatch_group_enter(deleteFilesTask);
+    for(NSString *name in names) {
+        dispatch_group_enter(deleteFilesTask);
+        [self deleteRemoteFileWithName:name completionHandler:^(BOOL success, NSUInteger bytesAvailable, NSError * _Nullable error) {
+            if(!success) {
+                failedDeletes[name] = error;
+            }
+            dispatch_group_leave(deleteFilesTask);
+        }];
+    }
+    dispatch_group_leave(deleteFilesTask);
+
+    // Wait for all files to be deleted
+    dispatch_group_notify(deleteFilesTask, dispatch_get_main_queue(), ^{
+        if (completionHandler == nil) { return; }
+        if (failedDeletes.count > 0) {
+            return completionHandler([NSError sdl_fileManager_unableToDeleteError:failedDeletes]);
+        }
+        return completionHandler(nil);
+    });
+}
 
 #pragma mark - Uploading
 
+- (void)uploadFiles:(NSArray<SDLFile *> *)files completionHandler:(nullable SDLFileManagerMultiUploadCompletionHandler)completionHandler {
+    [self uploadFiles:files progressHandler:nil completionHandler:completionHandler];
+}
+
+- (void)uploadFiles:(NSArray<SDLFile *> *)files progressHandler:(nullable SDLFileManagerMultiUploadProgressHandler)progressHandler completionHandler:(nullable SDLFileManagerMultiUploadCompletionHandler)completionHandler {
+    if (files.count == 0) {
+        return completionHandler([NSError sdl_fileManager_noFilesError]);
```

Comment:
We should throw an exception here instead. This is a programmer error, not one that occurs "accidentally" in production.